### ISSUE TITLE
[Bug] Triply-redeclared types defining methods annotated by self-referential relative forward references fail inscrutably (much like this title)

### ISSUE THREAD

**Issue by iamrecursion** [beartype/beartype#365]
I am sorry to turn up to the bear's cave yet again, but I must beseech thee for assistance!

Recently I've been doing a lot of cryptographic research in IPython notebooks, and use the bear's wisdom to keep myself sane by enforcing more complex type hints than mypy will cope with (see #350 for the levels of dark magic I would _like_ to be able to invoke, but such arcane incantations are not involved here). 

Recently I've run up against some oddness where repeated evaluations of a notebook cell exhibiting a certain pattern will flop between successfully passing the bear's scrutiny and getting swiped at by its claws. In particular, this only happens within a live IPython session where I am re-evaluating a cell that requests the bear's scrutiny.

### So What Is This Pattern?

Yeah, good point. I should probably provide some _code_ that shows the issue.

```python
from beartype import beartype

@beartype
class MyClass:
    attribute: int
    
    def __init__(self, attr: int) -> None:
        self.attribute = attr
        
    @classmethod
    def new_incremented(cls, attr: int) -> "MyClass":
        return MyClass(attr + 1)
        
instance = MyClass.new_incremented(10)
```

Any guesses as to what the 50%-of-the-time complaint is?

```
BeartypeCallHintReturnViolation: Class method __main__.MyClass.new_incremented() return <__main__.MyClass object at 0x107ecc990> violates type hint 'MyClass', as <class "__main__.MyClass"> <__main__.MyClass object at 0x107ecc990> not instance of <class "__main__.MyClass">.
```

Seems a bit odd, though I can imagine why it's happening. 

### Theories from the Typing Codices

Now I've obviously not spent anywhere near enough time in the internals of beartype to really understand the deep magicks it uses, but I can see a potential cause for this. Jupyter (or really IPython) keeps things around in an interactive session, and I think that this may well be the crux of the issue:

1. When we first execute the notebook, `MyClass` gets defined successfully, and the deferred `"MyClass"` type hint on `new_incremented` is not resolved until that class definition is done! Awesome!
2. However, when we execute that cell again, `MyClass` is already defined. I can easily see that beartype might resolve that type hint as soon as it can in that case, for performance!
3. But therein lies the rub! By the time we get to calling `new_incremented`, the `MyClass` it is being called on is a _new_ definition of the class (and hence returning an instance of that new definition), and not the one that was hanging around from the previous execution!
4. Boom! Type hint violation! We did indeed return an instance of `MyClass` that was not an instance of `MyClass`! Oops!

Frankly I have no clue how one might resolve this. It might be very simple, or fiendishly complicated, but it is not for me to tell the great bear its business!

### Blood for the Bear God

No bug report can come without ritual sacrifice (I think), so here's all the information that should be necessary, as well as a small repro that you can use!

**Beartype Version:** 0.18.2 (as mentioned in the title)  
**Python Version:** 3.11.8  
**Jupyter Version:** 1.0.0

I will note that this happens with `from __future__ import annotations` as well, rather than the string-y type hint declaration.

All you need to run this is a notebook execution environment with IPython and beartype installed, so I've set one up [here](https://github.com/iamrecursion/beartype_pls) that you can use as follows:

1. Clone the repo.
2. `poetry install --no-root`
3. Launch the notebook in your host of choice and make sure it is pointed at the venv listed by `poetry env list`.
4. Run the whole notebook once. It will complete without errors.
5. Run it twice more. It will give some variation of the above error.

**Comment by iamrecursion** [comment]
I wonder if this might be fixed by any of the machinery used to fix #364?

**Comment by leycec** [comment]
**tl;dr:** *Please globally replace all relative forward references of the form `"{class_name}"` with the `typing.Self` type hint. See below for long-winded blowhard discussion.*

**Yes!** It's my favourite bleeding-edge typonista, @iamrecursion. I now ramble incoherently. 

> Recently I've been doing a lot of **cryptographic research**...

*...heh.* ~~Black~~ white-hat NSA stuff, huh? Oh, Gods. This is like [Matt Damon's infamous "Good Will Hunting" NSA interview scene](https://www.youtube.com/watch?v=mJHvSp9AKYg) all over again, isn't it? Pretty sure that's where I almost ended up when I took the military's cryptography test in high school. But... *why?* Why did I do that!? Poverty of both wallet and imagination is the answer. The Army head-hunted me for literally *years* after I took that test. Thankfully, I instead chose to live in a rotting cabin in the woods with coywolves and beavers. **Score 1:** @leycec.

I now inexplicably want a quantum computer – whatever that is. :smile: 

> use the bear's wisdom

Sadly, the bear is only as wise as its creator. These are clearly dire times.

> enforcing more complex type hints than mypy will cope with

**Hah!** I always knew mypy couldn't cope with the harsh reality. Now, we have official confirmation.

*...wait.* Even @beartype doesn't do what you want yet. Nobody's coping with the harsh reality of your codebase, huh? Cryptography. My head hurts just thinking about post-quantum. It's probably because I have no hair. If I still had hair, surely everything would be easier like when I still had youth. :smiling_face_with_tear: 

*...wait.* What are we even chatbotting about here? Wasn't there an issue somewhere around here...

```python
BeartypeCallHintReturnViolation: Class method __main__.MyClass.new_incremented() return
<__main__.MyClass object at 0x107ecc990> violates type hint 'MyClass', as <class "__main__.MyClass">
<__main__.MyClass object at 0x107ecc990> not instance of <class "__main__.MyClass">.
```

Oh. Right. I sigh mournfully in the black despair of a windy Friday evening. The cat is wheezing. Will nothing go right?

## Jupyter + Forward References = :boom: 

**Oh, Gods.** Non-determinism in Jupyter, huh? Your educated guess is uncomfortably educated. It's almost like *you* should be authoring the @beartype codebase instead of me. Please! Help! Author the @beartype codebase instead of me!

I jest, of course. I jest badly. I'm no longer funny at all.

So, you're right about everything. Forward references are a perpetual pain point (P<sup>3</sup>) for runtime type-checkers. Doubly so for forward references in frequently re-evaluated Jupyter cells. Each time you re-evaluate a cell defining a class, you redefine that class as a new object with a unique identity. Although the recently deceased and newly resurrected classes share the same names, they're otherwise distinct types that basically share *nothing* in common from @beartype's admittedly autistic low-level perspective. As an autist, I can say things like that.

For efficiency, @beartype caches previously instantiated forward reference proxy objects as well as the user-defined classes those proxies resolved to. Almost certainly, the problem lives here.

Even so, @beartype should be doing the right deterministic thing. Due to prior issues like this, @beartype already:

1. Internally detects when you're redefining a previously defined class decorated by the `@beartype` decorator.
2. Responds by clearing all internal type-checking caches – .

I double-checked just now, because I no longer trust my younger self. Indeed, @beartype *is* doing both of those things. Even unit tests agree. I'm stumped like a suspicious mound of [*Sandkings*](https://forwearemany.wordpress.com/wp-content/uploads/2009/04/sandkings.pdf) in your backyard. For example, you'd think that something like the following would exhibit your exact issue...

*...wait!* That's it. By generalizing your original example, we obtain this shambolic horror:

```python
from beartype import beartype

for _ in range(3):
    @beartype
    class MyClass:
        attribute: int
        
        def __init__(self, attr: int) -> None:
            self.attribute = attr
            
        @classmethod
        def new_incremented(cls, attr: int) -> "MyClass":
            return MyClass(attr + 1)
            
    instance = MyClass.new_incremented(10)
```

...which deterministically raises your exception:

```python
beartype.roar.BeartypeCallHintReturnViolation: Class method __main__.MyClass.new_incremented() return
<__main__.MyClass object at 0x7f4557cf4e60> violates type hint 'MyClass', as
<class "__main__.MyClass"> <__main__.MyClass object at 0x7f4557cf4e60> not instance of
<class "__main__.MyClass">.
```

Oddly, it has to be **3** iterations. **2** is too few. Let's escalate this issue to official bugdom. If you can reproduce it, you can kill it.

And what is this...

> ## Blood for the Bear God

**W40K** memes in my issue tracker!? Bear-Nurgle approves before spewing sickening honey everywhere.

## Piously Plead for Workarounds and Bear-Nurgle Shall Provide

You're in luck. It's Friday and I'm feeling magnanimous like a bloated royal from the 1100's. If you wouldn't mind trivially refactoring your code in a bothersome way that will annoy you to no end, @beartype will happily satisfy your code's every desire to work. Specifically, please globally replace all relative forward references of the form `"{class_name}"` with the `typing.Self` type hint ala:

```python
from beartype import beartype
from typing import Self

@beartype
class MyClass:
    attribute: int
    
    def __init__(self, attr: int) -> None:
        self.attribute = attr
        
    @classmethod
    def new_incremented(cls, attr: int) -> Self:
        return MyClass(attr + 1)
        
instance = MyClass.new_incremented(10)
```

I can confirm that type-checks deterministically across 10,000 repeat declarations of the same class. Of course, that will probably fail on the 10,001th iteration. (If you don't test it, it's broken.)

Indeed, globally replacing all relative forward references of the form `"{class_name}"` with the `typing.Self` type hint is a good idea *anyway* irrespective of this issue. For various reasons that reduce to "@leycec is waving his noodly appendages," type-checking `typing.Self` is a *lot* faster (and robuster) than type-checking relative forward references.

@beartype: *it hardly ever works, but when it does, it works.*

**Comment by leycec** [comment]
**Resolved by 1204de437e40ba.** Somebody, please feed @beartype coins the size of the world. Thank you. :coin: :earth_americas: → :bear:

**Comment by iamrecursion** [comment]
I would like to preface all of this long-winded babbling with a hearty _thanks_ for such a rapid fix @leycec! You are truly doing the bear's work here, and "bleeding edge" is indeed right it seems!

> tl;dr: Please globally replace all relative forward references of the form "{class_name}" with the typing.Self type hint. See below for long-winded blowhard discussion.

Unfortunately this causes MyPy to balk! Either it is not as wise as our dear bear, or something else is going on! 

```python
beartype_pls.ipynb:cell_1:14: error: Incompatible return value type (got "MyClass", expected "Self")  [return-value]
```

Now this _is_ curious! Beartype is always happy, but MyPy is not! Let's give ourselves something concrete to chew on!

```python
from typing import Self
from beartype import beartype

@beartype
class MyClass:
    attribute: int
    
    def __init__(self, attr: int) -> None:
        self.attribute = attr
        
    @classmethod
    def new_incremented(cls, attr: int) -> Self:
        return MyClass(attr + 1)

    @classmethod
    def mypy_happy(cls, attr: int) -> Self:
        return cls(attr + 1)
        
instance = MyClass.new_incremented(10)
```

Would you believe that it doesn't think that explicitly naming the class in question satisfies the `Self` type annotation? I totally would. 

Based on my reading of the `Self` [typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#self), both the bear and the python are unfortunately correct. The bear sees at runtime that we have not broken anything, while MyPy can only deal with the static nature to know that `cls` is the correct class instance! 

Thankfully it is simple enough to construct the instance using `cls(...)` instead of explicitly naming things. More robust too I'd wager!

Phew!

--- 

> For various reasons that reduce to "@leycec is waving his noodly appendages," type-checking typing.Self is a lot faster (and robuster) than type-checking relative forward references.

I had known about the speed differential, but had previously intended that this class method did indeed return an instance of the _base_ type, rather than the derived type. That means that the explicit naming would be necessary! Thankfully you are correct in this case and my assuptions had changed!

The bear's wisdom indeed!!

---

> It's almost like you should be authoring the https://github.com/beartype codebase instead of me. Please! Help! Author the https://github.com/beartype codebase instead of me!

I can say nothing to doing it _instead_ of you, but I genuinely am up for helping to the best of my ability, starting with #350 if you think it doable within the codebase! Type checkers are a terrible twisted kind of home for me, and while I have little experience with _runtime_ ones I am sure I could do _some_ good. Perhaps!
